crates/editor/src/display_map.rs

```diff
@@ -1514,6 +1540,228 @@ impl ToDisplayPoint for Anchor {
     }
 }
 
+impl SemanticTokenView {
+    pub fn new(
+        buffer_id: BufferId,
+        multibuffer: &MultiBuffer,
+        lsp: &SemanticTokens,
+        legend: &lsp::SemanticTokensLegend,
+        cx: &App,
+    ) -> Option<SemanticTokenView> {
+        let Some(buffer) = multibuffer.buffer(buffer_id) else {
+            return None;
+        };
+        let buffer = buffer.read(cx);
+
+        let stylizer = SemanticTokenStylizer::new(legend);
+
+        let mut tokens = lsp
+            .tokens()
+            .filter_map(|token| {
+                let start = text::Unclipped(text::PointUtf16::new(token.line, token.start));
+                let (start_offset, end_offset) = point_offset_to_offsets(
+                    buffer.clip_point_utf16(start, Bias::Left),
+                    text::OffsetUtf16(token.length as usize),
+                    &buffer,
+                );
+
+                Some(MultibufferSemanticToken {
+                    range: start_offset..end_offset,
+                    style: stylizer.convert(
+                        cx.theme().syntax(),
+                        token.token_type,
+                        token.token_modifiers,
+                    )?,
+                    lsp_type: token.token_type,
+                    lsp_modifiers: token.token_modifiers,
+                })
+            })
+            .collect::<Vec<_>>();
+
+        // These should be sorted, but we rely on it for binary searching, so let's be sure.
```

Comment:
NIT: given `.collect::<Vec<_>>();`, it's also possible to use `binary_search_by` for insertion into a `Vec::new()` as a "single" operation instead of two.

Reply:
Collect and sort is likely cheaper given the sheer amount of tokens we might be working with. bin search + insert will have to move a lot of elements around otherwise